BUG: PyArray_CountNonzero does not check for exceptions

ahaldane opened the pull request:
Fixes #9295 

(reopen of #9296 which I could not reopen for some reason).

@eric-wieser I made the check for errors occur inside the loop.

ahaldane committed:
BUG: PyArray_CountNonzero does not check for exceptions

Fixes #9295

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -2093,13 +2093,17 @@ PyArray_CountNonzero(PyArrayObject *self)
         while (count--) {
-            if (nonzero(data, self)) {
+            npy_bool n = nonzero(data, self);
+            if (n == -1) {
```
This isn't even possible, is it?

ahaldane replied:
Ugh you're right, I should have caught that. `npy_bool` is `unsigned char`.

I was trying to avoid calling `PyErr_Occurred` every time through the loop. My timing tests suggested that gives a 40% slowdown.

eric-wieser replied:
In which case, the answer to "Should this be in the loop" is "no" ;)

ahaldane replied:
Well, correctness beats performance. If we are required by Python to check inside the loop, we may have to accept the slowdown (or find some workaround such as a special loop for Object dtypes).

Particularly in the case of Object types, I think the call to `PyObject_IsTrue` can enter back to python code, so I wouldn't want to leave stray error states.

ahaldane replied:
I think a lot of the ufunc inner loops had a similar problem which was solved by adding a `needs_api` variable to signify whether the error check could be delayed until after the loop.

ahaldane commented:
Fixed up. This version seems to have almost no performance impact on my system, based on the following test:
```python
[1]: a = np.arange(10000000)
[2]: %timeit a.nonzero()
```

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -2081,11 +2083,18 @@ PyArray_CountNonzero(PyArrayObject *self)
 
+    if (dtype->type_num == NPY_OBJECT || dtype->type_num == NPY_VOID ||
+            dtype->type_num >= NPY_NTYPES ||
+            PyDataType_FLAGCHK(dtype, NPY_NEEDS_PYAPI)) {
```
Should this last `||` be an `&&`? Can `PyDataType_FLAGCHK(...)` be true without meeting any of the earlier conditions?

ahaldane replied:
After studying, you are right, I think I just need to check the flag. Object dtypes have `NPY_OBJECT_DTYPE_FLAGS` set which includes the PYAPI flag, and voids have it set too. For user-defined types it is up to the implementor.

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -2073,6 +2073,8 @@ PyArray_CountNonzero(PyArrayObject *self)
     char *data;
     npy_intp stride, count;
     npy_intp nonzero_count = 0;
+    int needs_api = 0;
```
Might be tempted to use `npy_bool`

ahaldane replied:
It appears to traditionally be `int` in numpy, according to grep. So I'll leave it.

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -2081,11 +2083,16 @@ PyArray_CountNonzero(PyArrayObject *self)
 
+    if (PyDataType_FLAGCHK(dtype, NPY_NEEDS_PYAPI)) {
+        needs_api = 1;
+    }
```
Or: `needs_api = PyDataType_FLAGCHK(dtype, NPY_NEEDS_PYAPI)`

ahaldane replied:
Good, will do.

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -2096,10 +2103,13 @@ PyArray_CountNonzero(PyArrayObject *self)
         }
 
-        return nonzero_count;
+        return PyErr_Occurred() ? -1 : nonzero_count;
```
What's the rationale for checking in the loop only if an API is needed, but checking outside always?

Are non-`NPY_NEEDS_PYAPI` types allowed to throw errors? If not, then I think this should be more like

```
if(PyErr_Occured()) throw_SystemError("Must set that flag thing");
```

ahaldane replied:
I think currently none of the numpy types throw any error in nonzero. But I think in principle they could, or maybe a user-defined type might while not needing the PYAPI.

So I think it is more future-friendly to pass on potential errors rather than throwing our own.

Also, this way the code behaves like the ufunc api loops do.

eric-wieser replied:
Isn't `PyErr_SetString` part of the PYAPI? So how would it set the error?

But if this is how the ufuncs do it, then sure, let's fix that another time

eric-wieser replied:
The pattern I'm finding is `return (needs_api && PyErr_Occurred()) ? -1 : 0;` though - so are you sure that:

* We need to catch the error mid-loop, rather than wait till the end
* We need to catch anything at all if `!needs_api`

seberg replied:
It seems plausible to me that nonzero returns a special value on error, might want to use that as well. Ufunc machinery checks for error in the outer loop, the inner loop function should return on error though.

eric-wieser replied:
`nonzero` has return type `npy_bool`, which [usually](http://thedailywtf.com/articles/What_Is_Truth_0x3f_) has no special values left to use

ahaldane replied:
In `lowlevel_strided_loops.c.src` I see a couple places which break inner loops on `(needs_api && PyErr_Occurred())`. So I do think we need to catch errors mid loop. Furthermore, as noted in another comment above, if we don't break the loop we are likely to go back into python code on the next looparound with an error set.

Also, to restate my last comment, currently we do not actually need to catch anything if `!needs_api`, but perhaps in the future we might.

And @seberg, nonzero can't return special values because it has to return `npy_bool`, which should only really be 0 or 1, even if technically that is a char.

seberg replied:
I  half expected that cast, but it is silly, because python convention is to return an error indictor value, and that is also the case here (and nonzero is not much more then that function):

> int PyObject_IsTrue(PyObject *o)
> Returns 1 if the object o is considered to be true, and 0 otherwise. This is equivalent to the Python expression not not o. On failure, return -1.

Ignoring that means you do a more expensive check far more often. @ahaldane of course you need to break the inner loop, once the python error is set, you need to either clear it or stop doing python stuff, continuing the loop would do python stuff.

About "in the future we might to catch something", well, its a pretty safe thing to do, though I doubt there is much use here. If you have a type that can easily throw errors, it will "need_api". Though yes, we have a bit of problems with indicating errors for types that don't typically give errors....

charris commented:
Put the backport milestone off to 1.3.2.

ahaldane commented:
Just to be clear, did anyone request changes? My impression was that at worst the `PyError_Occurred` after the loop was unnecessary, but harmless, so I can leave it in. (I still think it could conceivably prevent future bugs).

eric-wieser commented:
I'd like to change things to `return (needs_api && PyErr_Occurred()) ? -1 : 0;`

> Also, to restate my last comment, currently we do not actually need to catch anything if !needs_api, but perhaps in the future we might.

But right now, we don't catch this in any of the ufuncs either. So let's be consistent, and only catch exceptions that should have been thrown in the first place.

If it turns out we do need to catch this, then we'll need to do so in all of these places anyway - and it'd be clearer if they all need the same fix

seberg commented:
Why put an error check anyway? If the lines above use a plain return statement, there is no plausible way for the code to reach an error state at that point? In any case I tend to agree with Eric. If the only reason is some (possibly far fetched) future change then it is probably just making the code more confusing in the now.

ahaldane commented:
I still disagree with that, because I am trying to account for the case where there is an error when `needs_api` is False. In other words, in this PR I want to account for two possibilities:

 1. `needs_api` is True. Then if there is an error I want to break the loop immediately and return an error, since each iteration of the loop may enter python code. This requires that I do an expensive check to `PyErr_Occurred` every time through the loop. The way I have it currently, we just immediately return from the function if an error occurs.
 2. `needs_api` is False. Then I want to avoid doing any expensive check inside the loop. If an error occurs I will let the loop continue as usual, and then just check for an error at the end.

In other words, I don't see the point of writing `return (needs_api && PyErr_Occurred()) ? -1 : nonzero;` at the end of the loop. Either `needs_api` is True and en error occurred, in which case we never reach this line. Or, `needs_api` if False and an error occurred, and we incorrecctly return success.

ahaldane commented:
Also, I think I may have confused the issue by referring to ufunc code too haphazardly above.

Let me consider two different kinds of loops in numpy: 1. Ufunc loops, 2. stransfer/copy loops

In summary, the ufunc loops don't check for `needs_api`, and I think nonzero is more like a ufunc than like a stransfter loop. 

More in detail:

1. The ufunc loops don't actually check for `needs_api` in their inner loops (`loops.c.src`). The only cases ufuncs can have an error are: Integers to a negative power, some datetime loops, and most of the Object loops. In these cases a Python Error is set, and the loop breaks. Control is passed back to a function in `ufunc_object.c`, which checks for just `PyErr_Occurred()` (*not* for `needs_api`). Eg, when you run `a = np.array([-1,2,3,4]); np.power(a, -1, where=(a<0))` it does this.  `needs_api` is mostly only used to control threading of the Npy_Iter.

One exception is in `ufunc_at` which *does* check for `needs_api` in a loop, and after it. I think that is an error though, since in principle a ufunc could have an error without `needs_api` being True, as the integer-power case demonstrates. But it's an error that never matters in current practice because, coincidentally, you can't use `np.power.at`.

2. The strided copy loops (`lowlevel_strided_loops.c.src`) check for `npy_api` and break inside their inner loops.  Then, after stransfer is called, the calling code often does `return (needs_api && PyErr_Occurred()) ? -1 : 0;`, which makes sense since errors can only occur if `needs_api` was true, unlike in the ufunc case.

ahaldane commented:
Incidentally, I noticed that the main example for how to use the iterator API in (the docs)[https://docs.scipy.org/doc/numpy/reference/c-api.iterator.html] is actually `PyArray_CountNonzero`, which we just found is buggy. 

I should probably update that doc page too.

ahaldane commented:
I'm going to leave the code in the docs alone, since it is already tweaked for simplicity anyway.

ahaldane commented:
I'll remove the error check though, I'm happy with leaving simpler, less confusing code we can fix if a problem turns up in the future.

Thanks for bearing with my argumentativeness on this one :)

eric-wieser commented on numpy/core/src/multiarray/sequence.c:
```diff
@@ -72,19 +72,26 @@ array_any_nonzero(PyArrayObject *arr)
     }
     Py_DECREF(it);
-    return anyTRUE;
+    return PyErr_Occurred() ? -1 : anyTRUE;
```
This should be `(needs_api && PyErr_Occured())`, to match the `if`.

I think it'd be clearer to just `DECREF(it); return -1;` instead of the `break` above though

ahaldane replied:
All right, done.

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -2145,11 +2152,15 @@ PyArray_CountNonzero(PyArrayObject *self)
             }
+            if (needs_api && PyErr_Occurred()) {
+                goto finish;
+            }
```
I think this should be `needs_api = NpyIter_IterationNeedsAPI(iter)`, for consistency with `NPY_BEGIN_THREADS_NDITER`

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -2081,11 +2083,13 @@ PyArray_CountNonzero(PyArrayObject *self)
                         PyArray_DIMS(self), PyArray_STRIDES(self));
     }
 
+    needs_api = PyDataType_FLAGCHK(dtype, NPY_NEEDS_PYAPI);
```
Move this inside the trivially iterable, and then use `NpyIter_IterationNeedsAPI` in the other case

ahaldane committed:
MAINT: array_any_nonzero now uses new iterator API

ahaldane commented:
It appears that `array_any_nonzero` is using the "old" iterator API, which made it difficult to make your suggested change.

So I've gone ahead and upgraded it to use the "new" iterator API, just like `PyArray_CountNonzero`.

eric-wieser commented:
My suggested change referred only to the use of the new api anyway, so I don't think the old api is a concern. But I suppose I can get behind using the new api in as many places as possible.

Is there a performance impact behind using the new api?

ahaldane commented:
It appears to be faster now:

Setup is `x = np.arange(100000)`

Numpy master:
```python
>>> %timeit 99999 in x
1000 loops, best of 3: 344 µs per loop

```

This PR:
```python
>>> %timeit 99999 in x
1000 loops, best of 3: 263 µs per loop
```

(numbers are consistent across multiple runs)

eric-wieser commented on numpy/core/src/multiarray/sequence.c:
```diff
@@ -69,22 +69,67 @@ NPY_NO_EXPORT PySequenceMethods array_as_sequence = {
+
+    NpyIter_Deallocate(iter);
+
+    return PyErr_Occurred() ? -1 : anyTRUE;
```
This still seems questionable to me, but I realize you've copied it from the old code, so perhaps it needn't be addressed here.

I'd be happier with renaming `npy_bool anyTRUE` to `int ret`, and then doing `ret = -1; goto finish;` in the error case

eric-wieser commented on numpy/core/src/multiarray/sequence.c:
```diff
@@ -69,22 +69,67 @@ NPY_NO_EXPORT PySequenceMethods array_as_sequence = {
-    PyArrayIterObject *it;
+    PyArray_NonzeroFunc *nonzero;
+    npy_intp stride, count;
+    char *data;
```
These three declarations would be tidier inside the `do` loop

ahaldane commented:
Updated.

ahaldane commented on numpy/core/src/multiarray/sequence.c:
```diff
@@ -69,22 +69,66 @@ NPY_NO_EXPORT PySequenceMethods array_as_sequence = {
+        while (count--) {
+            if (nonzero(data, arr)) {
+                ret = NPY_TRUE;
+                goto finish;
```
Also, I fixed a bug that this line said "break" in the last version you reviewed.

eric-wieser replied:
Are you testing on 1d arrays? On 2d arrays, having a `break` here ought to be a lot _slower_

ahaldane replied:
Yes, 1d.  Also my mistake,  I think had another process using cpus when I ran the test the 2nd time, now the speeds are as expected (a bit faster in this PR than master, also with goto).

eric-wieser commented:
Looks great, thanks. Can I get you to update `PyArray_CountNonzero` to not use `PyErr_Occured()` in the `return` statement, for consistency? I know you didn't touch it in this patch.

ahaldane commented:
Sure, done!

eric-wieser commented on numpy/core/tests/test_numeric.py:
```diff
@@ -1107,6 +1107,11 @@ def test_array_method(self):
+        # gh-9295
+        a = np.array([np.array([1, 2]), 3])
+        assert_raises(ValueError, np.nonzero, a)
+
```
Another useful test might be
```python
class BoolErrors:
    def __bool__(self):
        raise ValueError("Not allowed")

np.array([BoolErrors(), BoolErrors()]).nonzero()
```
which is similar to the test that came up for sorting objects with erroring comparators

juliantaylor commented on numpy/core/src/multiarray/sequence.c:
```diff
@@ -69,22 +69,66 @@ NPY_NO_EXPORT PySequenceMethods array_as_sequence = {
-    npy_intp counter;
-    PyArrayIterObject *it;
-    npy_bool anyTRUE = NPY_FALSE;
+    PyArray_NonzeroFunc *nonzero;
```
this function is only called from `array_contains` which should ensure the input is a boolean array. So we should just be able to use `count_boolean_trues`. While that function does not early abort it is so heavily optimized that it is probably still faster for many cases and its performance is irrelevant compare to the richcompare earlier

juliantaylor replied:
just doing PyArray_CountNonZero() > 0` gives these timings for 100000 int64 elements

```
original:
slow exit: 341
fast exit: 41us

using count_nonzero:
46.1us
46.1us
```

So slightly slower for fast exits but it removes some non-trivial code so I think it is worthwhile.
If we want to keep the early exit as is one could add a fast exit option to `count_boolean_trues`

ahaldane replied:
We could also call `PyArray_Any`, which seems even more literal. Then `array_contains` would be exactly equivalent to `(a == val).any()`.

`PyArray_Any` doesn't short-circuit either, but if anyone makes it do so in the future then we would get the speed benefit for free in `array_contains`.

juliantaylor commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -2081,21 +2083,25 @@ PyArray_CountNonzero(PyArrayObject *self)
 
     /* If it's a trivial one-dimensional loop, don't use an iterator */
     if (PyArray_TRIVIALLY_ITERABLE(self)) {
+        needs_api = PyDataType_FLAGCHK(dtype, NPY_NEEDS_PYAPI);
```
orthogonal to the issue, but this loop could release the GIL if `needs_api` is false, probably best via `NPY_BEGIN_THREADS_THRESHOLDED`

ahaldane committed:
MAINT: Simplify array_contains using PyArray_Any

ahaldane committed:
ENH: release GIL in CountNonzero trivial loop

ahaldane commented:
Updated following suggestions.
 * `PyArrayCountNonzero` now releases the GIL in trivial iterations if `needs_api` is false. I also split this into two branches, so now the `needs_api == False` case also skips the inner check for errors, making it just as fast as before.
 * `array_contains` now just calls `PyArray_Any`. This has no "fast break"/"short circuit" like there was before, but should still usually be faster. If we care about speed we should upgrade PyArray_Any to short-circuit anyway.
 * Added an extra test using custom object type.

charris commented:
Is it agreed that this is ready?

eric-wieser commented on numpy/core/src/multiarray/sequence.c:
```diff
@@ -32,16 +30,18 @@ array_contains(PyArrayObject *self, PyObject *el)
         return -1;
     }
-    ret = array_any_nonzero((PyArrayObject *)res);
+    any = PyArray_Any((PyArrayObject *)res, -1, NULL);
```
Is this the same? Doesn't that do `bool(arr.any(axis=-1))`, when before it did `arr.any()`?

ahaldane commented:
Important catch @eric-wieser! Fixed.

For some reason I thought `-1` was the sentinel value in the C-api, not `NPY_MAXDIMS`.

eric-wieser commented:
> For some reason I thought -1 was the sentinel value in the C-api, not NPY_MAXDIMS.

Can this be added to the documentation in `doc\source\reference\c-api.array.rst`?

ahaldane commented:
It is documented [here](https://docs.scipy.org/doc/numpy-1.13.0/reference/c-api.array.html#calculation) in the "Tip" at the start.

charris commented:
LGTM. I'll merge this in a bit if no one complains.

charris commented:
Thanks Allan.
